onnxruntime/core/optimizer/gemm_transpose_fusion.cc

```diff
@@ -28,14 +28,26 @@ Status GemmTransposeFusion::Apply(Graph& graph, Node& node, RewriteRuleEffect& m
   if (A_node_ptr != nullptr && A_node_ptr->OpType() == "Transpose") {
     Node& A_node = *graph.GetNode(A_node_ptr->Index());
     transA = !transA;
-    nodes_to_remove.push_back(A_node);
+    if (A_node.GetOutputEdgesCount() > 1) {
+      // remove the edge between the Transpose and Gemm nodes
```

Comment:
are we checking anywhere else that all the consumers are Gemms?
Also will we modify the transA/B for all the consumer Gemms here itself?
What if the other consumer is a MatMul which can also be fused?

Reply:
We don't need to check all consumers to be Gemm, it could that some comsumers are non-gemm nodes. Thus we process one gemm at a time. The graph will detect change, and loop back again if there's another transpose/gemm to be fused.
Other case like MatMul should be taken care by other transformers

Reply:
We still need to trigger this optimization only for All Gemm consumers case because if we don't, then there will be a double cost of transpose- one in the leftover op transpose, and the other within the Gemm kernel as transposes within the kernel are not always free.